File: frontend/app/lib/notes.ts
/**
 * Browser-side notes + highlights client (NF-11).
 *
 * Thin helpers the notes screens and the reader's capture popover use to drive
 * whole-Markdown notes and book highlights against FastAPI *through the same-origin
 * Next.js proxy* (`/api/...`, ADR-017) — never cross-origin. The HttpOnly session
 * cookie rides along automatically (`credentials: "same-origin"`), so this code
 * never reads or holds the session token. The reads (list, detail, backlinks)
 * carry no token; the state-changing writes (update, delete, capture) echo
 * the CSRF token (from `/api/auth/me`) in `X-CSRF-Token` (AD-007), mirroring
 * `quiz.ts`.
 *
 * A note is only ever born from a passage: `captureHighlight` is the one creation
 * helper here, because a note without a reading anchor can no longer exist. There
 * is deliberately no helper for a bare title.
 *
 * FastAPI remains authoritative for auth, ownership, anchor resolution, and the
 * body cap; these helpers just carry inputs in and surface the note/summary/
 * backlink/error out. The response types mirror the backend views in `web/notes.py`
 * exactly so the two never drift. Non-OK responses are surfaced as a typed
 * `NoteError` whose `kind` lets callers branch on the documented failures — a
 * stale capture (409, the reader tells the user to reload) and an over-cap body
 * (422, the editor flags the length) — without matching on message strings.
 */

/** A note's book citation, mirroring the backend `NoteAnchorView` (NF-10). */
export type NoteAnchor = {
  id: string;
  source_id: string;
  source_title: string;
  anchor: string;
  section_path: string[];
  block_ordinal: number | null;
  start_offset: number | null;
  end_offset: number | null;
  quote_exact: string;
  quote_prefix: string;
  quote_suffix: string;
  status: string;
};

/** The note-detail read model, mirroring the backend `NoteDetailView`. */
export type NoteDetail = {
  id: string;
  title: string;
  body_markdown: string;
  tags: string[];
  anchors: NoteAnchor[];
  created_at: string;
  updated_at: string;
};

/**
 * The passage a book-scoped note row came from, mirroring the backend
 * `NoteRowAnchorView` (WSN-02): the note's earliest anchor on the book being
 * listed. `page` is the book-global page number the server derived — never
 * recomputed here — and is null when the anchor no longer resolves, so an orphaned
 * row shows its quote with no page rather than a fabricated one.
 */
export type NoteRowAnchor = {
  anchor: string;
  section_title: string;
  section_path: string[];
  quote_exact: string;
  status: string;
  page: number | null;
};

/**
 * One row in the notes list, mirroring the backend `NoteSummaryView` (NF-13).
 *
 * `anchor` is carried only when the list is scoped to one book; the cross-book
 * list has no single book to represent a note by, so it is null there.
 */
export type NoteSummary = {
  id: string;
  title: string;
  tags: string[];
  anchor_statuses: string[];
  anchor: NoteRowAnchor | null;
  created_at: string;
  updated_at: string;
};

/** One inbound wikilink for the backlinks panel, mirroring `BacklinkView`. */
export type Backlink = {
  note_id: string;
  title: string;
};

/** The update body, mirroring the backend `NoteWriteRequest`. */
export type NoteWrite = {
  title: string;
  body_markdown?: string;
  tags?: string[];
};

/**
 * The highlight-capture body, mirroring the backend `CaptureRequest`. The
 * selection payload (`quote_exact` + 32-char `quote_prefix`/`quote_suffix`
 * context, resolved server-side against the section blocks) rides alongside the
 * new note's fields; an empty `body_markdown` yields a bare highlight.
 */
export type CaptureHighlight = {
  anchor: string;
  quote_exact: string;
  quote_prefix?: string;
  quote_suffix?: string;
  title: string;
  body_markdown?: string;
  tags?: string[];
};

/**
 * The documented failure the caller can branch on. `stale_capture` (409) means
 * the served evidence no longer matches the section (a mid-flight re-ingest) so
 * the reader should reload; `body_too_long` (422) means the note body exceeds the
 * cap so the editor should flag it; every other non-OK response collapses to
 * `unknown`.
 */
export type NoteErrorKind = "stale_capture" | "body_too_long" | "unknown";

/**
 * A non-OK notes response, carrying the mapped `kind` and the backend `detail`
 * (or a readable fallback) as its message. Callers match on `kind`, not the
 * message, so copy changes never break control flow.
 */
export class NoteError extends Error {
  readonly kind: NoteErrorKind;
  readonly status: number;

  constructor(kind: NoteErrorKind, status: number, message: string) {
    super(message);
    this.name = "NoteError";
    this.kind = kind;
    this.status = status;
  }
}

/**
 * List the caller's notes (newest-edited first), optionally filtered to one tag
 * (matched case-insensitively) and/or narrowed to one book. Under `sourceId` the
 * list holds only notes anchored to that book — a note anchored to it more than
 * once still appears once — and every row carries the passage it came from; a
 * source the caller does not own collapses to a 404 like every other source read.
 * Read-only GET, so no CSRF token. On a non-OK response the backend `detail` is
 * surfaced via `NoteError`.
 */
export async function listNotes(
  { tag, sourceId }: { tag?: string; sourceId?: string } = {},
  fetchImpl: typeof fetch = fetch,
): Promise<NoteSummary[]> {
  const params: string[] = [];
  if (tag !== undefined) {
    params.push(`tag=${encodeURIComponent(tag)}`);
  }
  if (sourceId !== undefined) {
    params.push(`source_id=${encodeURIComponent(sourceId)}`);
  }
  const query = params.length ? `?${params.join("&")}` : "";
  const res = await fetchImpl(`/api/notes${query}`, {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toNoteError(res, "Could not load your notes.");
  }
  return (await res.json()) as NoteSummary[];
}

/**
 * Fetch one owned note's detail (200). Read-only GET, so no CSRF token. A missing
 * or non-owned note collapses to a 404 surfaced via `NoteError`.
 */
export async function getNote(
  noteId: string,
  fetchImpl: typeof fetch = fetch,
): Promise<NoteDetail> {
  const res = await fetchImpl(`/api/notes/${noteId}`, {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toNoteError(res, "Could not load that note.");
  }
  return (await res.json()) as NoteDetail;
}

/**
 * Update an owned note and rewrite its derived indexes (200). State-changing, so
 * it carries the CSRF token in `X-CSRF-Token`. An over-cap body surfaces as a
 * `NoteError` with kind `body_too_long` (422).
 */
export async function updateNote(
  noteId: string,
  body: NoteWrite,
  csrfToken: string,
  fetchImpl: typeof fetch = fetch,
): Promise<NoteDetail> {
  const res = await fetchImpl(`/api/notes/${noteId}`, {
    method: "PATCH",
    credentials: "same-origin",
    headers: { "content-type": "application/json", "X-CSRF-Token": csrfToken },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    throw await toNoteError(res, "Could not save the note.");
  }
  return (await res.json()) as NoteDetail;
}

/**
 * Delete an owned note (204, no body). State-changing, so it carries the CSRF
 * token in `X-CSRF-Token`. A missing or non-owned note collapses to a 404
 * surfaced via `NoteError`.
 */
export async function deleteNote(
  noteId: string,
  csrfToken: string,
  fetchImpl: typeof fetch = fetch,
): Promise<void> {
  const res = await fetchImpl(`/api/notes/${noteId}`, {
    method: "DELETE",
    credentials: "same-origin",
    headers: { "X-CSRF-Token": csrfToken },
  });
  if (!res.ok) {
    throw await toNoteError(res, "Could not delete the note.");
  }
}

/**
 * Fetch the notes whose wikilinks resolve to an owned note (200). Read-only GET,
 * so no CSRF token. Owner-scoped like `getNote` (missing/non-owned → 404).
 */
export async function getBacklinks(
  noteId: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Backlink[]> {
  const res = await fetchImpl(`/api/notes/${noteId}/backlinks`, {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toNoteError(res, "Could not load the backlinks.");
  }
  return (await res.json()) as Backlink[];
}

/**
 * Capture a highlight from the reader: create a note + one book anchor atomically
 * (201). State-changing, so it carries the CSRF token in `X-CSRF-Token`. When the
 * served evidence no longer matches the section (a mid-flight re-ingest) the call
 * surfaces a `NoteError` with kind `stale_capture` (409); an over-cap body
 * surfaces kind `body_too_long` (422).
 */
export async function captureHighlight(
  sourceId: string,
  body: CaptureHighlight,
  csrfToken: string,
  fetchImpl: typeof fetch = fetch,
): Promise<NoteDetail> {
  const res = await fetchImpl(`/api/sources/${sourceId}/highlights`, {
    method: "POST",
    credentials: "same-origin",
    headers: { "content-type": "application/json", "X-CSRF-Token": csrfToken },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    throw await toNoteError(res, "Could not capture the highlight.");
  }
  return (await res.json()) as NoteDetail;
}

/**
 * Build a typed `NoteError` from a non-OK response. The status maps to a `kind`
 * (409 → `stale_capture`, 422 → `body_too_long`, else `unknown`) and the backend
 * `detail` string becomes the message; a 422 detail arrives as a list of
 * validation errors, not a string, so those fall back to the readable message
 * rather than a stringified list (mirroring `toQuizError`).
 */
async function toNoteError(res: Response, fallback: string): Promise<NoteError> {
  const kind: NoteErrorKind =
    res.status === 409
      ? "stale_capture"
      : res.status === 422
        ? "body_too_long"
        : "unknown";
  try {
    const body = (await res.json()) as { detail?: unknown };
    const message = typeof body.detail === "string" ? body.detail : fallback;
    return new NoteError(kind, res.status, message);
  } catch {
    return new NoteError(kind, res.status, fallback);
  }
}

File: backend/app/infrastructure/web/notes.py
"""Notes + highlights router — capture, organize, and read notes (Cycle E).

Thin FastAPI adapter over the framework-free notes services (assembled in
``dependencies``). A signed-in owner creates whole-Markdown notes, lists/filters
them by tag, reads/edits/deletes one, reads a note's backlinks, and captures a
highlight from the reader (create a note + one book anchor atomically). Every note
path is one request-scoped transaction, so no commit-then-enqueue orchestration is
needed here (unlike the deck/ingestion start paths).

Application errors are translated to HTTP by the global handlers
(``NoteNotFound`` → 404, ``NoteBodyTooLong`` → 422, ``StaleCaptureTarget`` → 409,
``SourceNotFound`` → 404, ``CorpusNotFound`` → 404).

Contract (also consumed by the Next.js proxy):
- ``POST   /api/notes`` → 201 note; auth + CSRF/Origin + limit.
- ``GET    /api/notes`` → 200 note summaries (optional ``?tag=`` filter); auth.
- ``GET    /api/notes/{id}`` → 200 note detail; auth.
- ``PATCH  /api/notes/{id}`` → 200 updated note detail; auth + CSRF/Origin + limit.
- ``DELETE /api/notes/{id}`` → 204; auth + CSRF/Origin + limit.
- ``GET    /api/notes/{id}/backlinks`` → 200 inbound links; auth.
- ``POST   /api/sources/{source_id}/highlights`` → 201 note; auth + CSRF/Origin +
  limit.

NF-09 lists an optional capture payload alongside ``POST /api/notes``. The reader's
capture path is served by the dedicated ``POST /api/sources/{source_id}/highlights``
route (the ``CaptureHighlight`` use case), which carries the source in the path and
the selection payload in the body; ``POST /api/notes`` stays a plain note create
(``CreateNote``, whose signature has no capture fields), so capture is not duplicated
across two shapes. This matches the design's architecture diagram (CRUD vs capture).
"""

from __future__ import annotations

from collections.abc import Callable
from contextlib import AbstractContextManager
from datetime import datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status
from pydantic import BaseModel, Field
from sqlalchemy import Connection

from app.application.notes import (
    CaptureHighlight,
    DeleteNote,
    GetBacklinks,
    GetNote,
    ListNotes,
)
from app.application.reading import ListSourceHighlights
from app.domain.entities import (
    Backlink,
    NoteAnchor,
    NoteSummary,
    NoteView,
    SourceHighlight,
    User,
)
from app.domain.ports import NoteIndexEnqueuer
from app.infrastructure.web.csrf import enforce_csrf, enforce_origin
from app.infrastructure.web.dependencies import (
    build_create_note,
    build_update_note,
    get_authenticated_user,
    get_capture_highlight,
    get_delete_note,
    get_get_backlinks,
    get_get_note,
    get_list_notes,
    get_list_source_highlights,
    get_note_index_enqueuer,
    get_note_uow,
)
from app.infrastructure.web.rate_limit import rate_limit_notes

router = APIRouter(tags=["notes"])

# Create/update commit in this UoW factory before the after-commit embed enqueue
# (AD-016), mirroring the ingestion/deck start paths; tests override both.
NoteUowFactory = Annotated[
    Callable[[], AbstractContextManager[Connection]], Depends(get_note_uow)
]
NoteEnqueuer = Annotated[NoteIndexEnqueuer, Depends(get_note_index_enqueuer)]


# --- Request bodies ------------------------------------------------------------


class NoteWriteRequest(BaseModel):
    """Create/update body (NF-05): title plus an optional Markdown body and tags.

    An empty body is allowed (a note may be just a title or, after capture, a bare
    quote card). Tags are normalized (lowercased/deduped) by the use case; the body
    cap is enforced there too (``NoteBodyTooLong`` → 422).
    """

    title: str
    body_markdown: str = ""
    tags: list[str] = Field(default_factory=list)


class CaptureRequest(BaseModel):
    """Highlight-capture body (NF-06): the selection payload plus the new note's fields.

    ``anchor`` addresses the reader section; ``quote_exact`` (+ 32-char
    ``quote_prefix``/``quote_suffix`` context) is the selection resolved server-side
    against the section's blocks. An empty ``body_markdown`` yields a bare highlight.
    """

    anchor: str
    quote_exact: str
    quote_prefix: str = ""
    quote_suffix: str = ""
    title: str
    body_markdown: str = ""
    tags: list[str] = Field(default_factory=list)


# --- Response views ------------------------------------------------------------


class NoteAnchorView(BaseModel):
    """A note's book citation (NF-10): the orphan-badge status + jump-back payload.

    Carries ``source_id``/``anchor``/``quote_exact`` so the reader can jump back to
    the passage (and re-place the highlight via the in-block offsets), plus the
    ``source_title`` snapshot and ``status`` so an orphaned anchor still renders from
    its quote after the source is gone.
    """

    id: UUID
    source_id: UUID
    source_title: str
    anchor: str
    section_path: list[str]
    block_ordinal: int | None
    start_offset: int | None
    end_offset: int | None
    quote_exact: str
    quote_prefix: str
    quote_suffix: str
    status: str

    @classmethod
    def from_anchor(cls, anchor: NoteAnchor) -> NoteAnchorView:
        return cls(
            id=anchor.id,
            source_id=anchor.source_id,
            source_title=anchor.source_title,
            anchor=anchor.anchor,
            section_path=list(anchor.section_path),
            block_ordinal=anchor.block_ordinal,
            start_offset=anchor.start_offset,
            end_offset=anchor.end_offset,
            quote_exact=anchor.quote_exact,
            quote_prefix=anchor.quote_prefix,
            quote_suffix=anchor.quote_suffix,
            status=anchor.status,
        )


class NoteDetailView(BaseModel):
    """The note-detail read model (NF-05/10): the note, its tags, and its anchors."""

    id: UUID
    title: str
    body_markdown: str
    tags: list[str]
    anchors: list[NoteAnchorView]
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_view(cls, view: NoteView) -> NoteDetailView:
        return cls(
            id=view.note.id,
            title=view.note.title,
            body_markdown=view.note.body_markdown,
            tags=list(view.tags),
            anchors=[NoteAnchorView.from_anchor(a) for a in view.anchors],
            created_at=view.note.created_at,
            updated_at=view.note.updated_at,
        )


class NoteSummaryView(BaseModel):
    """One row in the notes list (NF-13): the note with its tags and anchor statuses.

    ``anchor_statuses`` lets the list render active/stale/orphaned badges without
    loading the anchor payloads.
    """

    id: UUID
    title: str
    tags: list[str]
    anchor_statuses: list[str]
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_summary(cls, summary: NoteSummary) -> NoteSummaryView:
        return cls(
            id=summary.note.id,
            title=summary.note.title,
            tags=list(summary.tags),
            anchor_statuses=list(summary.anchor_statuses),
            created_at=summary.note.created_at,
            updated_at=summary.note.updated_at,
        )


class BacklinkView(BaseModel):
    """One inbound wikilink for the backlinks panel (NF-10/13): the linking note."""

    note_id: UUID
    title: str

    @classmethod
    def from_backlink(cls, backlink: Backlink) -> BacklinkView:
        return cls(note_id=backlink.note_id, title=backlink.title)


class SourceHighlightView(BaseModel):
    """One of the caller's highlights on a source, for inline reader painting (RD-28).

    The owning ``note_id`` plus the anchor's quote-with-context and ``status`` — the
    reader paints ``active`` quotes and ignores stale/orphaned ones (RD-29). The
    origin note's ``note_title`` and a ``has_body`` flag let the margin rail label each
    entry and tell a bare highlight from an annotated one without a second request
    (CAP-19); the painter ignores both.
    """

    note_id: UUID
    anchor: str
    quote_exact: str
    quote_prefix: str
    quote_suffix: str
    status: str
    note_title: str
    has_body: bool

    @classmethod
    def from_highlight(cls, highlight: SourceHighlight) -> SourceHighlightView:
        return cls(
            note_id=highlight.note_id,
            anchor=highlight.anchor,
            quote_exact=highlight.quote_exact,
            quote_prefix=highlight.quote_prefix,
            quote_suffix=highlight.quote_suffix,
            status=highlight.status,
            note_title=highlight.note_title,
            has_body=highlight.has_body,
        )


# --- Endpoints -----------------------------------------------------------------


@router.post(
    "/api/notes",
    status_code=status.HTTP_201_CREATED,
    dependencies=[
        Depends(rate_limit_notes),
        Depends(enforce_origin),
        Depends(enforce_csrf),
    ],
)
def create_note(
    user: Annotated[User, Depends(get_authenticated_user)],
    uow_factory: NoteUowFactory,
    enqueuer: NoteEnqueuer,
    body: NoteWriteRequest,
) -> NoteDetailView:
    """Create a whole-Markdown note for the caller and derive its indexes (201).

    ``CreateNote`` validates the body cap (``NoteBodyTooLong`` → 422) then persists
    the note and rebuilds its wikilink/tag indexes in one committed UoW; the note is
    then embedded asynchronously (only when it has a body to embed, NL-01), enqueued
    after commit so the worker reads a durable row (AD-016).
    """
    with uow_factory() as conn:
        view = build_create_note(conn)(
            user=user, title=body.title, body_markdown=body.body_markdown, tags=body.tags
        )
    if body.body_markdown:
        enqueuer.enqueue_embed(view.note.id)
    return NoteDetailView.from_view(view)


@router.get("/api/notes")
def list_notes(
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[ListNotes, Depends(get_list_notes)],
    tag: Annotated[str | None, Query()] = None,
) -> list[NoteSummaryView]:
    """Return the caller's notes (newest-edited first), optionally filtered by tag (200).

    ``tag`` is matched case-insensitively; every returned summary still lists all of
    its own tags.
    """
    summaries = service(user=user, tag=tag)
    return [NoteSummaryView.from_summary(s) for s in summaries]


@router.get("/api/notes/{note_id}")
def get_note(
    note_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[GetNote, Depends(get_get_note)],
) -> NoteDetailView:
    """Return an owned note's detail (200; 404).

    ``GetNote`` collapses a missing note and a non-owner to ``NoteNotFound`` → 404 so
    a note's existence is never disclosed.
    """
    return NoteDetailView.from_view(service(user=user, note_id=note_id))


@router.patch(
    "/api/notes/{note_id}",
    dependencies=[
        Depends(rate_limit_notes),
        Depends(enforce_origin),
        Depends(enforce_csrf),
    ],
)
def update_note(
    note_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    uow_factory: NoteUowFactory,
    enqueuer: NoteEnqueuer,
    body: NoteWriteRequest,
) -> NoteDetailView:
    """Update an owned note and rewrite its derived indexes (200; 404/422).

    Missing/non-owner → ``NoteNotFound`` → 404; over-cap body → ``NoteBodyTooLong`` →
    422. The wikilink/tag indexes are rebuilt from the new body in one committed UoW;
    the note is re-embedded asynchronously only when its body actually changed (a
    title/tags-only edit enqueues nothing, NL-01), after commit (AD-016).
    """
    with uow_factory() as conn:
        view, body_changed = build_update_note(conn)(
            user=user,
            note_id=note_id,
            title=body.title,
            body_markdown=body.body_markdown,
            tags=body.tags,
        )
    if body_changed:
        enqueuer.enqueue_embed(note_id)
    return NoteDetailView.from_view(view)


@router.delete(
    "/api/notes/{note_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[
        Depends(rate_limit_notes),
        Depends(enforce_origin),
        Depends(enforce_csrf),
    ],
)
def delete_note(
    note_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[DeleteNote, Depends(get_delete_note)],
) -> Response:
    """Delete an owned note (204; 404).

    Its anchors/tags/links cascade and inbound links from other notes are SET NULL;
    missing/non-owner → ``NoteNotFound`` → 404.
    """
    service(user=user, note_id=note_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/api/notes/{note_id}/backlinks")
def get_note_backlinks(
    note_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[GetBacklinks, Depends(get_get_backlinks)],
) -> list[BacklinkView]:
    """Return the notes whose wikilinks resolve to an owned note (200; 404).

    Owner-scoped like ``GetNote`` (missing/non-owner → ``NoteNotFound`` → 404).
    """
    backlinks = service(user=user, note_id=note_id)
    return [BacklinkView.from_backlink(b) for b in backlinks]


@router.post(
    "/api/sources/{source_id}/highlights",
    status_code=status.HTTP_201_CREATED,
    dependencies=[
        Depends(rate_limit_notes),
        Depends(enforce_origin),
        Depends(enforce_csrf),
    ],
)
def capture_highlight(
    source_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[CaptureHighlight, Depends(get_capture_highlight)],
    body: CaptureRequest,
) -> NoteDetailView:
    """Capture a highlight from the reader: create a note + one anchor atomically (201).

    ``CaptureHighlight`` authorizes the source (missing/non-owner → ``SourceNotFound``
    → 404), resolves the addressed section (unknown anchor → ``CorpusNotFound`` → 404),
    and binds the selection against its blocks — if the served evidence no longer
    matches (a mid-flight re-ingest) nothing is persisted and ``StaleCaptureTarget`` →
    409. Over-cap body → ``NoteBodyTooLong`` → 422.
    """
    view = service(
        user=user,
        source_id=source_id,
        anchor=body.anchor,
        quote_exact=body.quote_exact,
        quote_prefix=body.quote_prefix,
        quote_suffix=body.quote_suffix,
        title=body.title,
        body_markdown=body.body_markdown,
        tags=body.tags,
    )
    return NoteDetailView.from_view(view)


@router.get("/api/sources/{source_id}/highlights")
def get_source_highlights(
    source_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[ListSourceHighlights, Depends(get_list_source_highlights)],
) -> list[SourceHighlightView]:
    """Return the caller's highlights on an owned source (200); 404 missing/non-owner.

    Owner-scoped like the other source reads (``SourceNotFound`` → 404); every status is
    returned so the reader paints the ``active`` quotes and lists the rest.
    """
    return [
        SourceHighlightView.from_highlight(h)
        for h in service(user=user, source_id=source_id)
    ]
